### PR TITLE
CCS: ensure invalid storage objects are deleted

### LINKED ISSUE #364: Rotation of secrets fails due to "Secret xxx not found"
Rotation of secrets for `hashicorp/vault-secrets-operator:0.2.0` often fails showing the following errors in the logs: 

```
2023-09-18T07:21:46Z	INFO	Got deletion timestamp	{"controller": "vaultdynamicsecret", "controllerGroup": "secrets.hashicorp.com", "controllerKind": "VaultDynamicSecret", "VaultDynamicSecret": {"name":"database-rds","namespace":"infra"}, "namespace": "infra", "name": "database-rds", "reconcileID": "aa98f8e7-4f86-4c1a-a1bd-e96bab0f13d5", "obj": {"apiVersion": "secrets.hashicorp.com/v1beta1", "kind": "VaultDynamicSecret", "namespace": "infra", "name": "database-rds"}}
2023-09-18T07:21:46Z	INFO	Revoking lease for credential 	{"controller": "vaultdynamicsecret", "controllerGroup": "secrets.hashicorp.com", "controllerKind": "VaultDynamicSecret", "VaultDynamicSecret": {"name":"database-rds","namespace":"infra"}, "namespace": "infra", "name": "database-rds", "reconcileID": "aa98f8e7-4f86-4c1a-a1bd-e96bab0f13d5", "id": "management/rds/creds/eks-service-database-owner/Dy8esyd5is0VlN7vNT3hfsAc"}
{"level":"info","ts":"2023-09-18T07:21:46Z","logger":"clientCacheFactory","msg":"Setting up Vault Client for storage encryption","persist":true,"enforceEncryption":true,"cacheKey":"kubernetes-0ee175333e92d8a3e710da"}
2023-09-18T07:21:46Z	ERROR	cachingClientFactory	Failed to restore client from storage	{"controller": "vaultdynamicsecret", "controllerGroup": "secrets.hashicorp.com", "controllerKind": "VaultDynamicSecret", "VaultDynamicSecret": {"name":"database-rds","namespace":"infra"}, "namespace": "infra", "name": "database-rds", "reconcileID": "aa98f8e7-4f86-4c1a-a1bd-e96bab0f13d5", "error": "Secret \"vso-cc-kubernetes-3e6be83294f886a892cf48\" not found"}
github.com/hashicorp/vault-secrets-operator/internal/vault.(*cachingClientFactory).Get
	/home/runner/work/vault-secrets-operator/vault-secrets-operator/internal/vault/client_factory.go:309
github.com/hashicorp/vault-secrets-operator/controllers.(*VaultDynamicSecretReconciler).revokeLease
	/home/runner/work/vault-secrets-operator/vault-secrets-operator/controllers/vaultdynamicsecret_controller.go:469
github.com/hashicorp/vault-secrets-operator/controllers.(*VaultDynamicSecretReconciler).handleDeletion
	/home/runner/work/vault-secrets-operator/vault-secrets-operator/controllers/vaultdynamicsecret_controller.go:443
github.com/hashicorp/vault-secrets-operator/controllers.(*VaultDynamicSecretReconciler).Reconcile
	/home/runner/work/vault-secrets-operator/vault-secrets-operator/controllers/vaultdynamicsecret_controller.go:103
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
	/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:122
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:323
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:274
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:235
```
CachingClientFactory is setup with encryption: 

```
{"level":"info","ts":"2023-09-17T10:57:53Z","logger":"initCachingClientFactory","msg":"Initializing the CachingClientFactory"}
{"level":"info","ts":"2023-09-17T10:57:53Z","logger":"clientCacheFactory","msg":"Setting up Vault Client for storage encryption","persist":true,"enforceEncryption":true,"cacheKey":""}
2023-09-17T10:57:54Z	INFO	setup	Starting manager	{"clientCachePersistenceModel": "direct-encrypted", "clientCacheSize": 10000}
```

This is the VaultDynamicSecret:

```
apiVersion: secrets.hashicorp.com/v1beta1
kind: VaultDynamicSecret
metadata:
  namespace: infra
  name: database-rds
spec:
  vaultAuthRef: keycloak
  mount: management/rds
  path: creds/eks-service-database-owner
  revoke: true
  destination:
    create: true
    name: database-rds
  rolloutRestartTargets:
  - kind: StatefulSet
    name: database-mirror
```

Am I overseeing something?

### PR DESCRIPTION
**Pull request by benashz**
Previously, it was possible for invalid client cache storage objects to live forever. This could cause the ClientFactory to fail for valid Client requests. This was most noticeable on a call to RestoreAll() when VSO was first coming up.

Whenever an invalid cached client is found in storage it should be discarded and replaced with a valid client instance.

Other fixes:
- Add more granular metrics for tracking storage deletions which are performed during storage pruning.
- Increase test coverage for cache_storage
- Remove unused EncryptionRequiredError

Closes #364

### CODE REVIEW

**Review by swenson — APPROVED** [review]
LGTM

File: internal/vault/cache_storage.go [hunk 1]
```diff
@@ -49,7 +49,18 @@ const (
 	labelProviderNamespace    = "provider/namespace"
 )
 
-var EncryptionRequiredError = fmt.Errorf("encryption required")
+// commonMatchingLabels are included in all stored Clients. They are used for
+// nearly all client cache functions. Ideally they should never be altered,
+// since it could lead to orphaned cache entries and possibly other issues.
+var commonMatchingLabels = ctrlclient.MatchingLabels{
+	"app.kubernetes.io/name":       "vault-secrets-operator",
```
Comment (kschoche) [inline internal/vault/cache_storage.go:56]: Is this the right label, does the helm release name get appended to the app name?

**Review by kschoche — APPROVED** [review]
Looks good! Just had a question about the labels but otherwise looks great!

File: internal/vault/cache_storage.go (hunk 1 above)
Comment (benashz) [reply internal/vault/cache_storage.go:56]: Yes, these should be the right labels for the client cache secrets. They are Helm agnostic, and should survive a potential re-installation of VSO etc.

**Pull request merged**